### PR TITLE
Feat/tone instrument

### PR DESCRIPTION
**Pull request by lauraversocke**

### CODE REVIEW

File: src/features/instrument/Instrument.tsx [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+};
+const mediumVolume = new Volume(-15);
+const sampler = new Sampler({
+  E4: DogSample,
+}).chain(mediumVolume, Destination);
+const polySynthSquare = new PolySynth(Synth, {
+  oscillator: {
+    type: "square",
+  },
+}).chain(new Volume(-25), Destination);
```
Comment (lauraversocke) [inline src/features/instrument/Instrument.tsx:21]: Brace yourself for a long comment :D  I put much time on this and need your help to clarify it.
There are values `mediumVolume`, `sampler`, `polySynthSquare` which I wanted to initialize for my component - they do not change, so enough to initialize once. Those are instruments I am using in other places to play something. I found several ways doing same, all are working as expected, but want to clarify **which one any why suits here most**:

**1st way** - add simple as const variables inside Instrument
Note: I know such declaration of variables doesn't preserve same value between rerenderings, but in my case that was not an issue: 
  1) same value is always reset (**should I in general worry about the reassignment of constant (without any computation)?**)
  2) no re-renderings happen in my app (maybe its a bit weird for an app, but no UI updates ended in my musical app so far)
++doesn't cause re-rendering
  But, I was not sure, since here I am initializing not a primitive constant, but object, so was thinking maybe I need to use smth else as hooks, since that's what most commonly I saw in other code examples
```
  const mediumVolume = new Volume(-15);
  const sampler = new Sampler({
    E4: DogSample,
  }).chain(mediumVolume, Destination);

  const polySynthSquare = new PolySynth(Synth, {
    oscillator: {
      type: "square",
    },
  }).chain(new Volume(-25), Destination);
```

**2nd way** - use useState inside Instrument
--setting of initial values causes rerenderings, which are not needed
--I do not use 2nd returned param from `useState` because I do not need to reset the values, it feels I do not use full power from useState in this way :D 
```
  const [mediumVolume] = useState(new Volume(-15));
  const [sampler] = useState(new Sampler({
    E4: DogSample,
  }).chain(mediumVolume, Destination));
  const [polySynthSquare] = useState(new PolySynth(Synth, {
    oscillator: {
      type: "square",
    },
  }).chain(new Volume(-25), Destination));
```
  
**3rd way** - use useRef
++setting of initial values doesn't cause rerenderings
--I set variables to `.current`value, which feels like I intentionally make it not mutable and useRef isn't the right way here
  ```
  const mediumVolume = useRef(new Volume(-15)).current;
  const sampler = useRef(new Sampler({
    E4: DogSample,
  }).chain(mediumVolume, Destination)).current;
  const polySynthSquare = useRef(new PolySynth(Synth, {
    oscillator: {
      type: "square",
    },
  }).chain(new Volume(-25), Destination)).current;
```

**4th way** - values initialized once in useEffect having useState
-- doesnt make sense to initiate to null at first, since we know the needed value from the beginning
-- extra type checking later is needed in all cases
-- 2 useEffects are needed
 ```
 const [mediumVolume, setMediumVolume] = useState<Volume | null>(null);
  const [sampler, setSampler] = useState<Sampler | null>(null);
  const [polySynthSquare, setPolySynthSquare] = useState<PolySynth | null>();

  useEffect(() => {
    setMediumVolume(new Volume(-15));
    setPolySynthSquare(new PolySynth(Synth, {
      oscillator: {
        type: "square",
      },
    }).chain(new Volume(-25), Destination));
  }, []);

  useEffect(() => {
    if (mediumVolume) {
      setSampler(new Sampler({
        E4: DogSample,
      }).chain(mediumVolume, Destination));
    }
  }, [mediumVolume]);
```
  
**5th way** - values initialized once in useEffect having useRef
-- doesnt make sense to initialize null initially and later to update
-- requires extra type checking later is needed in all cases (on value and on `.current`)
  ```
  const mediumVolume = useRef<Volume | null>(null);
  const sampler = useRef<Sampler | null>(null);
  const polySynthSquare = useRef<PolySynth | null>();

  useEffect(() => {
    mediumVolume.current = new Volume(-15);

    sampler.current = new Sampler({
      E4: DogSample,
    }).chain(mediumVolume.current, Destination);

    polySynthSquare.current = new PolySynth(Synth, {
      oscillator: {
        type: "square",
      },
    }).chain(new Volume(-25), Destination);
  }, []);
  ```
**6th way** - the CURRENT one in PR - move outside the Instrument
 For me that makes sense the most, I just need to initialize the object values as constants, so all instance of Instrument components can share those. No need to cause rerenderng or deal with it.
 Just the part from [this article](https://thoughtspile.github.io/2021/10/25/useref-no-current/)
>... but now all instances of our component share the same value, _which is probably not what you want_

made me think is there a reason I wouldn't want it. I don't see, since do not update that value, I do not need to preserve anything among rerendering. **Do I miss anything, does my thinking ok?**

**TLDR** I wanted to initialize some values, firstly added as simple `const` variable (1st way). It felt not a react-ish way, `useState` usually described to keep the state of something, so I did (2nd way). Then researched how to initialized only once after a component is mounted  and `useEffect` was suggested (4th way). After that figured out that useState causes re-rendering which I do not need in my case and found the glory of `useRef` (3rd, 5th). Lastly, it felt like I'm trying to use a chainsaw instead of a butter knife to make a sandwich :D and returned to almost to 1st way, just moved outside the Instrument.

**Any mistakes/not considered parts in my thinking flow, plusses/minuses I noted for each way? What way would you suggest and why?**

File: src/features/instrument/Instrument.tsx [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+export const Instrument = () => {
+  // @TODO: separate to piaoa nd predefined isntruments
+
+  const isTransportStarted = useRef(false);
+  const melodyPart = useRef<Part | null>(null);
+  const bassPart = useRef<Part | null>(null);
+  const noise = useRef<Noise | null>(null);
```
Comment (lauraversocke) [inline src/features/instrument/Instrument.tsx:29]: These are the values I am using and changing in several functions. I struggled to decide how that's should be described.

Round 1 - initialized as `let` variables - [link to this code part of previous commit](https://github.com/laurasilvanavi/test-react-sounds-ts/pull/1/commits/6f6436c54d24fd69dd4ef53cbb776845dfb1db66#diff-dbc1ddb2229691475f99430fb4e92732592bebb1c3fd16f6d7f808098064864cR28-R31 )
--variables doesn't preserve value between re-renderings, in my case no rerendering happens, but it is very likely some rerendering would happen sooner or later in development of a component, so maybe better to have the initial setup ready

Round 2 - used useState - [link to this code part of previous commit](https://github.com/laurasilvanavi/test-react-sounds-ts/pull/1/commits/266106de6f85f2f1fc385b5c01767cf93db2b00a#diff-dbc1ddb2229691475f99430fb4e92732592bebb1c3fd16f6d7f808098064864cR18-R21)
-- initialization and value changes caused rendering, which wasn't needed

Round 3 - used useRef - the current one in PR
++ initialization or value changes doesnt cause rerendering
++ value changes are preserved between future re-renderings
-- it's hurts to write `.current` everywhere, feels like doing smth wrong

**What do you think about this one?**

File: src/App.tsx [hunk 3]
```diff
@@ -1,25 +1,48 @@
...
+  palette: {
+    mode: 'dark',
+    primary: {
+      main: common.white
+    },
+    secondary: teal,
+  },
+  components: {
+    MuiButton: {
+      styleOverrides: {
```
Comment (lauraversocke) [inline src/App.tsx:17]: 🤮

Does that look ok for you? 

Context: It's my first try to use some UI lib for React other than Polaris. I wanted to use some material design lib, `mui` appeared to be very popular in React world.
I found it so terrible to customize current components. And I'm using only buttons :D 
**It's ok if you are not familiar and don't want dig into, ignore this comment, but if you are - is customization supposed to be like that? That's all fine?** I'm considering to have a look into other libs later

It's so odd to set stylings via js/ts. But that's the only working and semi-okay-looking way I found, other options: 
-set styles in component's using `!important` everywhere
-set inline styles using still js/ts on every component using magic `sx` prop, couldn't avoid using once

I couldn't make to work:
-[overriding styles with class names](https://mui.com/customization/how-to-customize/#overriding-styles-with-class-names), not sure that's to be better due to weird-looking naming, TBH
-[other deprecated ways in which are still in docs](https://mui.com/styles/advanced/)

Comment (perrupa) [reply src/App.tsx:17]: Yeah I can't say I've used Mui before, so please take any of this with a grain of salt 😅 

From what I've seen working with themes in the past (minimal experience) the theme overides/customizations are generally reserved for modifying constants the theme uses (similar to polaris' `spacing` constants: loose, tight, etc... [exposed in polaris-tokens](https://github.com/shopify/polaris-tokens)) rather than for styling individual components.

[seems like mui calls these `theme config variables`](https://mui.com/customization/theming/#theme-configuration-variables)

If you wanted to modify a specific component, I think you may want to look into extending (cascading?) some additional styles using the same classnames that mui uses? Seems like they [have a few rules about CSS specificity you need to match](https://mui.com/customization/how-to-customize/#:~:text=Never%20style%20these%20state%20classes%27%20names%20directly%3A) for proper overriding.

Then your theme remains minimal, and you can have specific overides, per component, for each component in CSS.

> -set inline styles using still js/ts on every component using magic sx prop, couldn't avoid using once

The easier way might be to create one button component that does this, and then import that button for use everywhere. This kinda defeats the point of having theme-able components as now you have to modify and/or wrap their components with our own. Something we were trying to avoid by using a component library ⚰️

Overwriting their classnames w/ CSS may be the easiest way to override their styles, but I think maybe you might want to look into creating state variants for specific cases or custom components for edge-cases (I don't know if you want ALL your buttons to be `10em` wide 😅)

File: src/features/instrument/Instrument.tsx (hunk 1 above)
Comment (perrupa) [reply src/features/instrument/Instrument.tsx:21]: I think you've nailed it with #6 for your use case, it's by far the simplest and perfect for your app.

I think the only reason you'd want to look into any of the approaches would be if your app needed to be able to initialize/re-initialize the sound libraries based on the components' render lifecycle.

IF what you're looking for is a place to centralize your sampler/volume logic, then maybe what you're looking for is a central place to maintain that state, and a way to share/manipulate that state? If that's the case, you may be [in React.Context territory](https://github.com/laurasilvanavi/test-react-sounds-ts/pull/2#discussion_r821136779)

I'd also wonder if there's a different way to maintain state/share instances using tone.js, but unfortunately I'm a bit unfamiliar with that lib as well 😓

File: src/features/instrument/Instrument.tsx (hunk 2 above)
Comment (perrupa) [reply src/features/instrument/Instrument.tsx:29]: The main reason you would want to use `useState` over `useRef` would be if you wanted the component to rerender when that value changes.

If you don't need to rerender, useRef is fine. If you want to render changes (like maybe toggling play/pause or mute button states) then state is more appropriate.

I think what you really need is a central place for this song logic. Whether that's a single hook (vs a component) or a sharing the song parts + transport via a context.

File: src/features/instrument/Instrument.tsx [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+      noise.current.mute = !noise.current.mute;
+    } else {
+      noise.current = new Noise("pink").start();
+      const filter = new AutoFilter({
+        frequency: "8m",
+      }).chain(new Volume(-20), Destination);
+      noise.current.connect(filter);
+      filter.start();
+    }
+  }
```
Comment (perrupa) [inline src/features/instrument/Instrument.tsx:102]: Seems like we want either a `mutePart` function, or a `togglePart` function... Does the instrument need to be re-initialized each time it's played (vs start/stopping or mute/unmuting)?

File: src/features/instrument/Instrument.tsx [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+  // @TODO: check arent a beter way to write each thing now (always create new player?), also check updates in tone
+  function playSample(sampleName: string) {
+    new Player({
+      url: sampleName,
+      autostart: true,
+    }).chain(mediumVolume, Destination);
```
Comment (perrupa) [inline src/features/instrument/Instrument.tsx:109]: Seems like we could initialize the samples onload and then start/stop them vs creating new ones each time according to the docs?

https://tonejs.github.io/docs/14.7.77/Player#start

**Review by perrupa — COMMENTED** [review]
Some really great ideas in here, I think centralizing the app state is starting to happen here in the `Instrument` component though that is maybe not the best final place for it.

I would love to see it get lifted to a Context (or a hook) to better isolate the app logic.
